x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_rules_table_saved_state.ts

```diff
@@ -41,61 +37,55 @@ function validateState(
 ): [RulesTableSavedFilter, RulesTableSavedSorting, RulesTableUrlSavedPagination] {
   const [filterFromUrl] = validateNonExact(urlState, RulesTableSavedFilter);
   const [filterFromStorage] = validateNonExact(storageState, RulesTableSavedFilter);
-  const filter = { ...filterFromStorage, ...filterFromUrl };
+  const filter: RulesTableSavedFilter = {
+    searchTerm: filterFromUrl?.searchTerm ?? filterFromStorage?.searchTerm,
+    source: filterFromUrl?.source ?? filterFromStorage?.source,
+    tags: filterFromUrl?.tags ?? filterFromStorage?.tags,
+    enabled: filterFromUrl?.enabled ?? filterFromStorage?.enabled,
+  };
 
   const [sortingFromUrl] = validateNonExact(urlState, RulesTableSavedSorting);
   const [sortingFromStorage] = validateNonExact(storageState, RulesTableSavedSorting);
-  const sorting = { ...sortingFromStorage, ...sortingFromUrl };
+  const sorting = {
+    field: sortingFromUrl?.field ?? sortingFromStorage?.field,
+    order: sortingFromUrl?.order ?? sortingFromStorage?.order,
+  };
 
   const [paginationFromUrl] = validateNonExact(urlState, RulesTableUrlSavedPagination);
   const [paginationFromStorage] = validateNonExact(storageState, RulesTableStorageSavedPagination);
-  const pagination = { perPage: paginationFromStorage?.perPage, ...paginationFromUrl };
+  const pagination = {
+    page: paginationFromUrl?.page,
+    perPage: paginationFromUrl?.perPage ?? paginationFromStorage?.perPage,
+  };
+
+  if (
+    pagination.perPage &&
+    (pagination.perPage < 0 || pagination.perPage > RULES_TABLE_MAX_PAGE_SIZE)
+  ) {
+    delete pagination.perPage;
+  }
 
   return [filter, sorting, pagination];
 }
 
-export function useInitializeRulesTableSavedState(): void {
+export function useRulesTableSavedState(): {
```

Comment:
Nit: The code is easier to read if it's written with the "newspaper metaphor" in mind: it's readable from top to bottom, from higher-level abstractions or implementations to lower-level details. In this case, `useRulesTableSavedState` would be defined first, and then its helper functions.